### PR TITLE
Feat/1 package skeleton

### PR DESCRIPTION
**Pull request by AlexTheWizardL**
# Pull Request: Package Skeleton

## Description
This pull request adds initial structure and simple test to check connectivity. The key highlights of this pull request include:

- **GitHub Actions Workflow**: 
  - Implemented a new workflow `check-release` that triggers on push, pull_request, pull_request_target, and on a nightly schedule at 04:00 UTC. It runs tests using the latest Python container.
  - Added relevant environment variables and conditional checks to ensure the workflow functions as expected based on the event type and actor.

- **Git Ignore Update**:
  - Added `.DS_Store` to `.gitignore` to prevent system-related files from being tracked.

- **Documentation**:
  - Created a `README.md` file describing the `cashctrl_ledger` package, installation instructions, testing strategy, and package development guidelines.
  - Included steps for setting up a CashCtrl Pro account for API integration.
  
- **Package Development**:
  - Created a new Python package `cashctrl_ledger` with an `__init__.py` file and a module `ledger.py`.
  - Implemented the `CashCtrlLedger` class in `ledger.py` with various unimplemented methods intended for future expansion.
  - Added `setup.py` to configure the package installation and required dependencies.
  
- **Unit Tests**:
  - Created a test file `tests/test_connectivity.py` to validate basic connectivity and ensure data is not empty.
  - Implemented a fixture to create an instance of `CashCtrlLedger` for use in tests.
  - Added a test to check that the 'person/list.json' endpoint returns a response with non-empty data.

## ChatGPT Review
<details> <summary>Click for ChatGPT prompt</summary>

## Code Structure
The overall structure of the package aligns well with typical Python practices. The organization, with an `__init__.py` in the main package directory and the primary implementation in `ledger.py`, is clear.

- **Structure**: Consistent with standard Python practices for creating libraries.
- **File Organization**: Logical with separate files for package initialization and main class implementation.

### Suggestions
- **Method Naming**: Use descriptive names and ensure unimplemented methods have clear docstrings.
- **Method Grouping**: Group similar methods together to improve readability and maintainability.

## Code Style and PEP 8 Compliance
The code appears to follow Python's PEP 8 style guidelines, including indentation, spacing, and naming conventions.

- **PEP 8 Compliance**: Indentation and naming conventions seem correct.
- **Comments and Docstrings**: Helpful explanations are provided in several places.

### Suggestions
- **Type Checks**: Use `isinstance()` to ensure proper types for method parameters.
- **Trailing Whitespace**: Check for and remove trailing whitespace.
- **Import Organization**: Follow PEP 8 order (standard libraries, third-party, then internal imports).

## Unit Testing and CI/CD
The addition of unit tests and GitHub Actions workflow aligns with best practices for continuous integration/delivery.

- **Unit Tests**: `test_connectivity.py` verifies basic functionality and uses `pytest`.
- **GitHub Actions**: Workflow automates testing on various triggers and includes a nightly schedule.

### Suggestions
- **Test Coverage**: Increase to cover edge cases and error conditions.
- **Code Quality Checks**: Consider integrating `flake8` or `pylint` into the workflow for code linting.

## Documentation
The README provides a good overview of the package, its purpose, and setup instructions. The `CONTRIBUTING.md` file offers useful guidance for contributors.

- **README**: Includes installation instructions and testing strategy.
- **CONTRIBUTING.md**: Contains guidelines for contributing to the project.

### Suggestions
- **Code Examples**: Add to the README for common use cases and scenarios.
- **Method Documentation**: Provide detailed docstrings for expected input/output, even if methods are unimplemented.

## Final Thoughts
Overall, this pull request establishes a strong foundation for integrating with CashCtrl. The inclusion of unit tests, CI/CD setup, and thorough documentation is commendable.

Areas for improvement include expanding test coverage, refining package structure, and enhancing documentation with additional examples and method details. With these adjustments, the package will align more closely with Python community standards and support future development and contributions effectively.

</details>

### CODE REVIEW

File: .github/workflows/check-release.yml [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+      - name: Setup virtualenv
+        run: |
+          python -V  # Print out python version for debugging
+          pip install virtualenv pytest
+          virtualenv venv
+          . venv/bin/activate
+
+      - name: Install dependencies
+        run: pip install coverage pytest pandas .
```
Comment (lasuk) [inline .github/workflows/check-release.yml:46]: What does the dot at the end of the line stand for?

File: README.md [hunk 2]
```diff
@@ -1,2 +1,57 @@
...
+Tests are housed in the [cashctrl_ledger/tests](tests) directory and are automatically
+executed via GitHub Actions. This ensures that the code is tested after each
+commit, during pull requests, and on a daily schedule. We prefer pytest for its
+straightforward and readable syntax over the unittest package from the standard
+library.
+
+
+## Package Development and Contribution
+
+See [CONTRIBUTING.md](CONTRIBUTING.md) for:
```
Comment (lasuk) [inline README.md:50]: I suggest to link to CONTRIBUTING.md in the cashctrl_api package. Here's an adapted version:

```
See [CONTRIBUTING.md](macxred/cashctrl_api/CONTRIBUTING.md) for:
```

File: README.md [hunk 3]
```diff
@@ -1,2 +1,57 @@
...
+
+- Testing Strategy
```
Comment (lasuk) [inline README.md:52]: Testing strategy is included as a separate section in this README, no need to mention it as reference here.

File: setup.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
+from setuptools import setup, find_packages
+
+setup(
+    name="cashctrl_ledger",
+    version="v0.0.1",
+    description="Python package that implements the pyledger.LedgerEngine and integrates with cashctrl",
+    url='https://github.com/macxred/cashctrl_ledger',
+    author="Lukas Elmiger",
```
Comment (lasuk) [inline setup.py:8]: @AlexTheWizardL: Don't you want to be the package author?

File: tests/test_connectivity.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
+"""
+Unit tests for basic connectivity in CashCtrlLedger.
+These tests check that essential data is present and not empty.
+"""
+
+import pytest  # Using pytest framework for testing
+from cashctrl_ledger import CashCtrlLedger
+
+@pytest.fixture
```
Comment (lasuk) [inline tests/test_connectivity.py:9]: Is there any advantage of using a picture in this simple case?

Calling `CashCtrlLedger()` directly instead of the fixture would IMHO improve code readability.

File: tests/test_connectivity.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    """
+    Fixture to create an instance of CashCtrlLedger for use in tests.
+    """
+    return CashCtrlLedger()
+
+def test_data_is_not_empty(cashctrl_ledger):
+    """
+    Test that the 'person/list.json' endpoint returns a response with non-empty data.
+    """
+    # Act
```
Comment (lasuk) [inline tests/test_connectivity.py:20]: Comments 'Act' and 'Assert' are not adding any new information. I would drop them for brevity.

**Review by lasuk — COMMENTED** [review]
Thanks a lot. Looks great!

I've added a few minor comments, please implement or discuss further.

Also, now that CashCtrl credentials are available on the server, please ensure the test passes.

File: .github/workflows/check-release.yml (hunk 1 above)
Comment (AlexTheWizardL) [reply .github/workflows/check-release.yml:46]: The main goal here is to install the pyledger and cashctrl_api packages that should be installed from the git.

Let`s go through the steps:
1. pip install coverage pytest pandas (without "." for now)
2. then we have python setup.py develop. In this scenario, I receive such an error(screenshot linked)

The "." symbol simply installs all "install_requires" dependencies. after this error vanishes
Screenshot 2024-05-01 at 11 01 47

File: setup.py (hunk 4 above)
Comment (AlexTheWizardL) [reply setup.py:8]: Yeah, sure. Let's have `author="Lukas Elmiger, Oleksandr Stepanenko"`

File: .github/workflows/check-release.yml (hunk 1 above)
Comment (lasuk) [reply .github/workflows/check-release.yml:46]: Thanks. I just learned something :)

File: README.md (hunk 2 above)
Comment (AlexTheWizardL) [reply README.md]: Fixed

File: README.md (hunk 3 above)
Comment (AlexTheWizardL) [reply README.md]: Fixed

File: README.md (hunk 2 above)
Comment (lasuk) [reply README.md]: Relative links to other GitHub projects do not seem to work in this context. The link is broken (error 404). See:
https://github.com/macxred/cashctrl_ledger/tree/feat/1_package-skeleton?tab=readme-ov-file#package-development-and-contribution

We need to change to use an absolute path instead:
```
See [cashctrl_api /CONTRIBUTING.md](https://github.com/macxred/cashctrl_api/blob/main/CONTRIBUTING.md) for:
```

**Comment by lasuk** [comment]
Merging. Thanks for the great work!
